### DISCUSSION TITLE
Heat flux discrepancy for coupled MOOSE-THM simulation

### DISCUSSION THREAD

**Discussion by lewisgross1296** [idaholab/moose#28926] [Q&A Modules: Thermal Hydraulics] [answered]
I'm running a MOOSE-THM coupled simulation with the following scheme (MOOSE Heat Transfer Module is the main app)

- Run a 1-D THM solve on `timestep_begin` and use a [HeatTransferFromExternalAppTemperature1Phase](https://mooseframework.inl.gov/source/components/HeatTransferFromExternalAppTemperature1Phase.html) component for the heat transfer (the first `T_wall` is just the inlet temperature everywhere)
  -  Send fluid T and heat transfer coefficient (HTC) to MOOSE
- Then, run MOOSE  with a constant volumetric heat source, use a Robin ([CoupledConvectiveHeatFluxBC](https://mooseframework.inl.gov/source/bcs/CoupledConvectiveHeatFluxBC.html)) BC to solve for solid temperature
  - Compute a `NearestPointLayeredSideAverage` `T_wall` and send to each THM sub-application (this case has 31)
 
In the coupling scheme, I use a `ProjectionAux` to make a nodal `T_inf_nodal` variable to send back to MOOSE for the robin BC (as I was getting complaints about transferring between nodal and elemental variables). The temperature results converge to steady state (`steady_state_tolerance = 5e-5`) and seem reasonable.
 
As a verification excercise, I attempted to compute $q'' = HTC (T_{wall} - T_{inf})$ for each THM channel, but the heat flux looks a little wonky. My hypothesis is that mixing elemental and nodal variables is responsible for the produced $q_{THM}''$. I also produced a $q_{solid}''$ from the main app temperature field using a [DiffusionFluxAux](https://mooseframework.inl.gov/source/auxkernels/DiffusionFluxAux.html)  + `NearestPointLayeredSideAverage`. This heat flux looks much better compared to the THM one.

Here's how I computed $q_{THM}''$ (I've denoted which variables are nodal/elemental)
```
  [compute_q_wall] 
   type = ADConvectiveHeatFlux1PhaseAux
   Hw = Hw # elemental computed from Material objects (`ADWallFrictionChurchillMaterial`, `ADReynoldsNumberMaterial`, `ADPrandtlNumberMaterial`, the Dittus-Boelter correlation for `Nu` and `ADConvectiveHeatTransferCoefficientMaterial`)
   T = T # elemental solution results (main output of THM)
   T_wall = T_wall # nodal variable transferred from solid main app
   variable = q_wall # elemental AuxVariable 
  []
  ```
Now, I'll show the two results. From THM
![thm_q_wall_all_pipes](https://github.com/user-attachments/assets/4c13934b-3bdc-482b-a1af-c782dc1efe15)

From MOOSE
![solid_computed_heat_flux_all_pipes](https://github.com/user-attachments/assets/81677280-756d-4dd3-a260-6b4a2bfe7ab0)

Am I doing something wrong? I've tried all kinds of projections for the inputs above (e.g. making an elemental T_wall for use in the above kernel), and they all end up with strange bumps in the middle. 

I'm not sure the projections were helping or doing exactly the right thing in this case. Would definitely be interested to hear more about projecting between nodal and elemental variables for THM, since it seems to be the key to understanding the behavior here.

I can upload my input files, if desired.

**Comment by GiudGiud** [comment]
Hello

Sorry for the delay.
The projection would likely affect the heat flux, but here it seems there is a larger trend at play. 
What is at 1.2m in y in the simulation?

I think I am misunderstanding the setup here. 
The convective heat flux output by THM is of the form H(T-Twall) and the diffusive heat flux of the form D grad T_wall (right?).
Why do you expect both to match? Is diffusion the only mechanism on the solid side?

**Reply by lewisgross1296** [reply]
No worries, thanks for the response.

The strange thing is there's nothing special happening at 1.2 m in my simulation (at least not intentionally). The assembly is 2.4 m tall. If I understand correctly, since I'm using `n_elems = 48` for my `FlowChannel1Phase` component in THM, there would be a node at 1.2 m, but there's no physics occurring there that should make this point special.

There's a constant volumetric heat source on the solid side but only in the fuel. All the coolant channels occur in graphite, so diffusion is the only mechanism in those blocks. Here's my kernels in the solid app
```
[Kernels]
  [diffusion]
    type = HeatConduction
    variable = T
  []
  [time_T]
    type = HeatConductionTimeDerivative
    variable = T
  []
  [source]
    type = CoupledForce
    variable = T
    v = power_density
    block = 'fuel'
  []
[]
```
Since I'm using this BC on the solid side
```
[BCs]
  [robin_bc_on_fsi]
    type = CoupledConvectiveHeatFluxBC
    variable = T
    boundary = 'fluid_solid_interface'
    htc = 'HTC'
    T_infinity = 'T_inf'
  []
[]
```
I do expect $q'' = HTC (T - T_{inf})$ on the coolant channel boundary. Let me know if you would like to see the inputs, I can send a zip file since my repo is currently private. I also have a fair amount of other plots along the y axis (T_fluid, T_wall, HTC, etc.) if you are curious about those, I can post as well.

**Reply by GiudGiud** [reply]
> The assembly is 2.4 m tall. If I understand correctly, since I'm using n_elems = 48 for my FlowChannel1Phase component in THM, there would be a node at 1.2 m

on the solid mesh side, what kind of discretization do you have? If you have an odd number of elements, you could get something weird with a nearest-location approach, where the "wrong" node (both are valid choices, but maybe the decision on which one to map changes as this center node) gets mapped to the central node. 

> I also have a fair amount of other plots along the y axis (T_fluid, T_wall, HTC, etc.) if you are curious about those, I can post as well.

I think that's a good idea. We know q_wall is weird, but is it because of HTC, T_wall or T_fluid?
To find that out, run a loosely coupled simulation (no fixed point) so we do not see something like (wrong T_wall causes wrong T_fluid causes wrong HTC)

**Comment by lewisgross1296** [comment]
I'm using the same number of axial layers on the solid side as the number of elements in THM (`num_layers = 48` in my solid mesh extrusion step). I think the node y-coordinates should match exactly, but I see what you're saying about a potential nearest-location issue.

I'll give a quick rundown of each variable and its type. On the solid app, I've got
```
[Variables]
  [T] # LINEAR LAGRAGE (default)
  []
[]

[AuxVariables]
  # receive THM computed fluid temp
  [T_inf]
  []
  # receive THM computed heat transfer coefficient
  [HTC]
    family = MONOMIAL
    order = CONSTANT
  []
  # receive power density from OpenMC
  [power_density]
    family = MONOMIAL
    order = CONSTANT
    block = 'fuel'
  []
  # variable to compute heat flux
  [heat_flux_normal]
    family = MONOMIAL
    order = CONSTANT
  []
[]
```

THM creates `T` (elemental) and `T_wall` (nodal, `FIRST LAGRANGE`) and I add the AuxVars
```
[AuxVariables]
  [T_inf_nodal] # project elemental T onto this variable before transferring to solid app
  []
  [Hw] # materials are defined on elements
    family = MONOMIAL
    order = CONSTANT
  []
  [q_wall]
    family = MONOMIAL
    order = CONSTANT
  []
[]
```
And here are the VPPs output by THM
![thm_T_all_pipes](https://github.com/user-attachments/assets/f4ddbfb0-7adc-4014-a1e9-75c27022ec1e)
![thm_T_inf_nodal_all_pipes](https://github.com/user-attachments/assets/f8759047-dc9d-43ef-8784-35b6581994f9)
![thm_T_wall_all_pipes](https://github.com/user-attachments/assets/1bc54069-a02c-4ce5-8fc8-f9a832f98736)
![thm_HTC_all_pipes](https://github.com/user-attachments/assets/c4a3c8b7-dc96-4f52-bf22-e2c1667831c9)

Now that I'm looking at these, something is a bit off on the `T_inf_nodal` near the center (maybe caused by the projections). While it's `T` that is used in the `q_wall` computation, `T_inf_nodal` is used to compute `T_wall`. Perhaps the issue starts there?

> To find that out, run a loosely coupled simulation (no fixed point) so we do not see something like (wrong T_wall causes wrong T_fluid causes wrong HTC)

Do you mean run THM alone with some T_wall defined in my input?

**Reply by GiudGiud** [reply]
> Do you mean run THM alone with some T_wall defined in my input?

This jump could be having the consequences you see. Let's use a smooth shape for T inf nodal and see if it fixes it

**Reply by GiudGiud** [reply]
alternatively, could you change on the mesh to have a half-sized element at the start. Then the nodes & centroids would be aligned and you would not need to project?

**Reply by lewisgross1296** [reply]
Think I'm starting to see this is mostly a post processing / choice of number of layers in my `LayeredAverage` `[UserObjects]` issue.

I ran a simulation where instead of projecting an elemental `T` onto a nodal `T_inf_nodal`, I defined an analytical `T_inf_nodal` via an IC using the function $T_{inf~nodal} = 873.15 + y*(1100 - 873.15)/2.4$. As a sanity check, I loaded one of the `solid_out_thm<N>.e` into paraview and made a plot there, attached below

![paraview_T](https://github.com/user-attachments/assets/34873924-4007-4529-944c-ab0cbafa859a)

This shows that `T_inf_nodal` is a line (with no "bump" in the center), but `T_wall` sent from the solid still has something weird going on in the middle. For the same simulation, when I run my post processing scripts, I still see a bump in my `T_inf_nodal`. 
![T_inf_nodal_post_processed](https://github.com/user-attachments/assets/6f5bce83-5d23-4166-9a13-0a64fb470b92)

Thinking about it more, I believe the issue is with my choice of `num_layers` for the UO + VPP that outputs the CSV from which the above plot is generated.  My THM mesh has `num_layers_for_THM` (which is 48) elements. **Where I think I was going wrong was thinking that the layer averages for nodal quantities should have `num_layers_for_THM +1` layers, since there's one more node than element.**

What this ACTUALLY does  is average the `LINEAR LAGRANGE` variable over $\frac{N}{N+1}$ of the normal element height, which causes the interior layers to gets split over multiple mesh elements.

In my post processing script, I take those N+1 values and plot them at the node locations. This isn't technically correct, since they should probably be plotted at the centroid of each layer / probably is why there's a weird bump. Here's an image that shows this axial division for a simplified for a case with `N=4` layers (so 5 nodes). Basically, the purple points are the nodal values (`T_inf_nodal`) and occur at boundaries between elements and there would be a linear function between. As can be seen, using `N=5` layers to do the `LayeredAverage` isn't conformal with the mesh. 

![nodal_layer_av](https://github.com/user-attachments/assets/56dabbdf-23f1-4360-8173-c6744dd78eb8)

I'm not 100% sure if it averages the node values or the linear temperature function, but the key takeaway is the temp values are not being plotted at the correct locations (in my post processed plots) and the averaging domains are a bit off for nodal variables.



I think moving forward, I should just use N layers in all my `LayeredAverage` to be conformal with the mesh. I also think using Projections, despite this warning
```
*** Warning ***
/home/lgross/gcmr_playground/multiphysics/inner/MOOSE_THM_tweak/solid.i:240.3:
The following warning occurred in the Transfer 'T_inf_from_THM' of type MultiAppGeneralFieldNearestLocationTransfer.

Transfer is projecting from nearest-centroids to nodes. This is likely causing floating point indetermination in the results because multiple centroids are 'nearest' to a node. Please consider using a ProjectionAux to build a nodal source variable (for example linear Lagrange) before transferring
```
is probably not necessary / adds more variables (and confusion) / does the same thing the transfer between different variable types does (correct me if I'm wrong).


I also plan to make my THM `T_wall` variable elemental so it is more compatible with the UserObject transfer (aka a variable that receives N data points from solid)
```
  [ht_exit]
    type = HeatTransferFromExternalAppTemperature1Phase
    flow_channel = channel
    P_hf = '${fparse channel_diameter * pi}'
    T_ext = T_wall
    var_type = elemental 
  []
```

Waiting on a simulation with all these changes to run, but I'll post an update when it finishes.

**Reply by GiudGiud** [reply]
Do you mean N-1 layers? N-1 layers fit between N nodes (with the nodes at the edges of the layers, the layered average is the average of both nodal values)

Could you do N layers, but offset them so they are centered on the nodes?

**Reply by lewisgross1296** [reply, marked as answer]
So using `N` layers for all my `UserObjects` and no `ProjectionAux` (just transfer elemental `T` to solid app nodal variable `T_inf` to serve as `T_infinity` in my `CoupledConvectiveHeatFluxBC`) seems much better.

Here are the new plots generated from THM quantities (the T variable is essentially the same).

![thm_T_wall_all_pipes ](https://github.com/user-attachments/assets/18660d50-1af7-4679-8f87-f98b2c2539bc)

![thm_HTC_all_pipes](https://github.com/user-attachments/assets/b3d35b41-b273-4683-b52f-6e6422531680)

![thm_q_wall_all_pipes](https://github.com/user-attachments/assets/93b30746-07e3-490d-aaee-8bdb84031392)

There's still something a little off near the top of the system (can be seen in the HTC too), but overall this is much closer to the `q_wall` that the solid produces. I would like to test projection vs no projection at some point, but would need to wait for 1D nodal <-> elemental to be allowed again with projections.

**Reply by GiudGiud** [reply]
This looks so much better indeed. 
So removing the projection is the way to go?
or is the bulk of the fix really aligning the bins / using the right number?

The 1D projection change is in a  large PR of mine already. If you need it ASAP I can separate in another PR

**Reply by lewisgross1296** [reply]
I think the main improvement comes from using the correct number of layers in my UO. I am curious to see the change due to projection, but I can move forward without it for now. Care to ping me on the PR so I can watch it?
